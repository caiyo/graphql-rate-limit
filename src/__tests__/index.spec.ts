import { graphql } from 'graphql';
import gql from 'graphql-tag';
import {
  makeExecutableSchema,
  IResolverValidationOptions,
} from 'graphql-tools';
import { RateLimiterMemory, RateLimiterRes } from 'rate-limiter-flexible';
import { createRateLimitTypeDef, createRateLimitDirective } from '../index';

describe('createRateLimitTypeDef', () => {
  it('creates custom directive definition', () => {
    const directiveName = 'customRateLimit';
    const directiveTypeDef = createRateLimitTypeDef(directiveName);

    expect(directiveTypeDef.definitions[0].name.value).toBe(directiveName);
    expect(directiveTypeDef).toMatchSnapshot();
  });
});

describe('createRateLimitDirective', () => {
  const consume = jest.spyOn(RateLimiterMemory.prototype, 'consume');
  const resolvers = {
    Query: {
      books: () => [
        {
          title: 'A Game of Thrones',
          author: 'George R. R. Martin',
        },
        {
          title: 'The Hobbit',
          author: 'J. R. R. Tolkien',
        },
      ],
      quote: () =>
        'The future is something which everyone reaches at the rate of sixty minutes an hour, whatever he does, whoever he is. â€• C.S. Lewis',
    },
  };
  const resolverValidationOptions: IResolverValidationOptions = {
    allowResolversNotInSchema: true,
  };
  beforeEach(() => {
    consume.mockClear();
  });

  it('limits a field', async () => {
    const typeDefs = gql`
      type Query {
        quote: String @rateLimit
      }
    `;
    const schema = makeExecutableSchema({
      typeDefs: [createRateLimitTypeDef(), typeDefs],
      resolvers,
      resolverValidationOptions,
      schemaDirectives: {
        rateLimit: createRateLimitDirective(),
      },
    });

    const response = await graphql(schema, 'query { quote }');

    expect(response).toMatchSnapshot();
    expect(consume).toHaveBeenCalledTimes(1);
    expect(consume).toHaveBeenCalledWith('Query.quote');
  });
  it('limits a repeated field', async () => {
    const typeDefs = gql`
      type Query {
        books: [Book!]
      }
      type Book {
        title: String @rateLimit
      }
    `;
    const schema = makeExecutableSchema({
      typeDefs: [createRateLimitTypeDef(), typeDefs],
      resolvers,
      resolverValidationOptions,
      schemaDirectives: {
        rateLimit: createRateLimitDirective(),
      },
    });

    const response = await graphql(schema, 'query { books { title } }');

    expect(response).toMatchSnapshot();
    expect(consume).toHaveBeenCalledTimes(2);
    expect(consume).toHaveBeenCalledWith('Book.title');
  });
  it('limits an object', async () => {
    const typeDefs = gql`
      type Query @rateLimit {
        books: [Book!]
        quote: String
      }
      type Book {
        title: String
      }
    `;
    const schema = makeExecutableSchema({
      typeDefs: [createRateLimitTypeDef(), typeDefs],
      resolvers,
      resolverValidationOptions,
      schemaDirectives: {
        rateLimit: createRateLimitDirective(),
      },
    });

    const response = await graphql(schema, 'query { quote books { title } }');

    expect(response).toMatchSnapshot();
    expect(consume).toHaveBeenCalledTimes(2);
    expect(consume).toHaveBeenCalledWith('Query.quote');
    expect(consume).toHaveBeenCalledWith('Query.books');
  });
  it('raises limiter error', async () => {
    consume.mockImplementation(() => {
      throw new Error('Some error happened');
    });
    const typeDefs = gql`
      type Query {
        quote: String @rateLimit
      }
    `;
    const schema = makeExecutableSchema({
      typeDefs: [createRateLimitTypeDef(), typeDefs],
      resolvers,
      resolverValidationOptions,
      schemaDirectives: {
        rateLimit: createRateLimitDirective(),
      },
    });

    const response = await graphql(schema, 'query { quote }');

    expect(response).toMatchSnapshot();
    expect(consume).toHaveBeenCalledTimes(1);
    expect(consume).toHaveBeenCalledWith('Query.quote');
  });
  it('throttles on limit reached', async () => {
    consume
      .mockResolvedValueOnce({
        msBeforeNext: 250,
        remainingPoints: 0,
        consumedPoints: 1,
        isFirstInDuration: true,
      })
      .mockRejectedValue({
        msBeforeNext: 250,
        remainingPoints: 0,
        consumedPoints: 0,
        isFirstInDuration: false,
      });
    const typeDefs = gql`
      type Query {
        quote: String @rateLimit(limit: 1)
      }
    `;
    const schema = makeExecutableSchema({
      typeDefs: [createRateLimitTypeDef(), typeDefs],
      resolvers,
      resolverValidationOptions,
      schemaDirectives: {
        rateLimit: createRateLimitDirective(),
      },
    });

    const response = await graphql(
      schema,
      'query { firstQuote: quote secondQuote: quote }',
    );

    expect(response).toMatchSnapshot();
    expect(consume).toHaveBeenCalledTimes(2);
    expect(consume).toHaveBeenCalledWith('Query.quote');
  });
});
