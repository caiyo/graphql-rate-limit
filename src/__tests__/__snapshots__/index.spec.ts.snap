// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createRateLimitDirective limits a field 1`] = `
Object {
  "data": Object {
    "quote": "The future is something which everyone reaches at the rate of sixty minutes an hour, whatever he does, whoever he is. ― C.S. Lewis",
  },
}
`;

exports[`createRateLimitDirective limits an object 1`] = `
Object {
  "data": Object {
    "books": Array [
      Object {
        "title": "A Game of Thrones",
      },
      Object {
        "title": "The Hobbit",
      },
    ],
    "quote": "The future is something which everyone reaches at the rate of sixty minutes an hour, whatever he does, whoever he is. ― C.S. Lewis",
  },
}
`;

exports[`createRateLimitTypeDef creates custom directive definition 1`] = `
Object {
  "definitions": Array [
    Object {
      "arguments": Array [
        Object {
          "defaultValue": Object {
            "kind": "IntValue",
            "value": "60",
          },
          "description": Object {
            "block": true,
            "kind": "StringValue",
            "value": "Quantity that is allowed per period.",
          },
          "directives": Array [],
          "kind": "InputValueDefinition",
          "name": Object {
            "kind": "Name",
            "value": "limit",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "Int",
            },
          },
        },
        Object {
          "defaultValue": Object {
            "kind": "IntValue",
            "value": "60",
          },
          "description": Object {
            "block": true,
            "kind": "StringValue",
            "value": "Number of seconds before limit is reset.",
          },
          "directives": Array [],
          "kind": "InputValueDefinition",
          "name": Object {
            "kind": "Name",
            "value": "duration",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "Int",
            },
          },
        },
      ],
      "description": Object {
        "block": true,
        "kind": "StringValue",
        "value": "Controls the rate of traffic.",
      },
      "kind": "DirectiveDefinition",
      "locations": Array [
        Object {
          "kind": "Name",
          "value": "OBJECT",
        },
        Object {
          "kind": "Name",
          "value": "FIELD_DEFINITION",
        },
      ],
      "name": Object {
        "kind": "Name",
        "value": "customRateLimit",
      },
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 270,
    "start": 0,
  },
}
`;
